collector/jolokia_collector.go

```diff
@@ -0,0 +1,211 @@
+}
+
+type JolokiaRequest struct {
+	Attribute string `json:"attribute"`
+	MBean     string `json:"mbean"`
+	Path      string `json:"path"`
+}
+
+func (te *timeEpoch) UnmarshalJSON(b []byte) error {
+	var seconds int64
+	err := json.Unmarshal(b, &seconds)
+	if err == nil {
+		time := time.Unix(seconds, 0)
+		*te = timeEpoch(time)
+		return nil
+	} else {
+		return err
+	}
+}
+
+//Returns a new collector using the information extracted from the configfile
+func NewJolokiaCollector(collectorName string, configFile []byte, metricCountLimit int, containerHandler container.ContainerHandler, httpClient *http.Client) (*JolokiaCollector, error) {
+	var configInJSON Jolokia
+	err := json.Unmarshal(configFile, &configInJSON)
+	if err != nil {
+		return nil, err
+	}
+
+	configInJSON.Endpoint.configure(containerHandler, defaultEndpointConfig)
+
+	pollingFrequency, err := time.ParseDuration(configInJSON.PollingFrequency)
+	if err != nil {
+		glog.Warningf("Failed to properly parse the 'polling_frequency' values as a go time.Duration object. Setting to %vs.", defaultPollingFrequency.Seconds())
+		pollingFrequency = defaultPollingFrequency
+	}
+	if pollingFrequency < minpollingfrequency {
+		glog.Warningf("The polling frequency is lower than the minium, the poll frequency for %s has now been to the minium (%vs).", collectorName, minpollingfrequency.Seconds())
+		pollingFrequency = minpollingfrequency
+	}
+
+	if metricCountLimit < 0 {
+		return nil, fmt.Errorf("Metric count limit must be greater than 0")
+	}
+
+	if len(configInJSON.MetricsConfig) > metricCountLimit {
+		return nil, fmt.Errorf("Too many metrics defined: %d limit %d", len(configInJSON.MetricsConfig), metricCountLimit)
+	}
+
+	if !strings.HasSuffix(configInJSON.Endpoint.URL, "/") {
+		configInJSON.Endpoint.URL = configInJSON.Endpoint.URL + "/"
+	}
+
+	return &JolokiaCollector{
+		name:             collectorName,
+		configFile:       configInJSON,
+		metricCountLimit: metricCountLimit,
+		pollingFrequency: pollingFrequency,
+		httpClient:       httpClient,
+	}, nil
+}
+
+//Returns name of the collector
+func (collector *JolokiaCollector) Name() string {
+	return collector.name
+}
+
+func (collector *JolokiaCollector) configToSpec(config JolokiaMetricConfig) v1.MetricSpec {
+	return v1.MetricSpec{
+		Name:   config.Name,
+		Type:   config.MetricType,
+		Format: config.DataType,
+		Units:  config.Units,
+	}
+}
+
+func (collector *JolokiaCollector) GetSpec() []v1.MetricSpec {
+	specs := []v1.MetricSpec{}
+
+	for _, metricConfig := range collector.configFile.MetricsConfig {
+		spec := collector.configToSpec(metricConfig)
+		specs = append(specs, spec)
+	}
+	return specs
+}
+
+//Returns collected metrics and the next collection time of the collector
+func (collector *JolokiaCollector) Collect(metrics map[string][]v1.MetricVal) (time.Time, map[string][]v1.MetricVal, error) {
+	currentTime := time.Now()
+	nextCollectionTime := currentTime.Add(time.Duration(collector.pollingFrequency))
```

Comment:
Should the next collection time depend on whether the current collection succeeded? Like retry in a second if it fails, and retry in a minute if it succeeds?